ldap/servers/slapd/pw.c

```diff
@@ -735,6 +752,47 @@ update_pw_info(Slapi_PBlock *pb, char *old_pw)
     {
         pw_exp_date = NO_TIME;
         slapi_mods_add_string(&smods, LDAP_MOD_REPLACE, "pwdReset", "TRUE");
+        if ((pwpolicy->pw_otp_maxuse >= 0) ||
+            (pwpolicy->pw_otp_delay_expire_at >= 0) ||
+            (pwpolicy->pw_otp_delay_valid_from >= 0)) {
+            char *use_count, *expire_at, *valid_from;
+            time_t cur_time;
+            cur_time = slapi_current_utc_time();
+            slapi_log_err(SLAPI_LOG_TRACE,
+                          "update_pw_info",
+                          "OTP password reset by an admin, pwdOTPReset=TRUE\n");
+            slapi_mods_add_string(&smods, LDAP_MOD_REPLACE, "pwdOTPReset", "TRUE");
+
+            /* If useCount is -1, this means this limit is not enforced */
+            if (pwpolicy->pw_otp_maxuse >= 0) {
+                use_count = "0";
+            } else {
+                use_count = "-1";
+            }
+            slapi_log_err(SLAPI_LOG_TRACE, "update_pw_info", "pwdOTPUseCount = %s\n", use_count);
```

Comment:
One last comment - is it possible for someone to manually set these values? 

Consider I want to create a set of new accounts. I want them to be active from 9:00am local time, and the password expires at 5pm the same day. 

With the current "relative" generation of the expireat/validfrom it would be really hard to achieve this for bulk accounts.

So is it possible that I can set these values manually and it would be respected?

Reply:
'pwdOTP*' attributes are operational and direct updates are rejected with unwilling to perform. So only the server will set those values according to the policy.
The current PR propose to configure delays that are used to compute the validity window. So it is not possible to define precise time (e.g. April 21rst 2021 9:00am to April 21rst 2021 5pm).
Creating a bulk of accounts the validity window of the 1rst account will differ from the last account.
An option to do that would be to add in the password policy an attribute like 'passwordOTPValidityRef' that could be a fixed date. If this attribute is defined then the server takes it as reference instead of current time (during userpassword reset).

Reply:
@tbordaz I guarantee the number one request we will get from this feature, is exactly what I am requesting - the ability to set these values manually. As a past admin, it's exactly how I would approach it, and something I would need and take value from.

Reply:
@Firstyear, I agree that the use case is nice (an admin enabling a set of accounts that each one should be activited within a fixed period of time).
I propose to update the design and describe a possible option ('passwordOTPValidityRef') and if you agree to open a separated RFE.

The current use case is rather on demand reset. A user forgot his password, ask for a new temporary one (sent by email for example), then the use of the new password has some limitations (#auth and period of time).

Reply:
Well, I think adding more attributes isn't needed. Just make it so that these attributes can be set by an admin. That way:

* If the passwordOTPExpireAt and ValidFrom are NOT present, we generate them.
* If they are present, or they are set, we just use the new manually set values.

I think this is the simplest and clearest approach.

Reply:
Initially I thought that a change in the schema would allow that. I tried to drop 'NO-USER-MODIFICATION' and it works for DM. But it also works for a user having the rights to update entry. And I would not like to add an aci to deny updates of  those attributes.

So a solution would be a change in op_shared_is_allowed_attr path to allow DM to update those attributes even if they have 'NO-USER-MODIFICATION'.
I will work on a new patch with that solution.

Reply:
But if the user can modify the entry they can already affect nsAccountLock or the userPassword. So ACI is the correct way to manage this here .... I'm not really seeing the issue?

Reply:
Sorry my concern was invalid. This is fine defining as modifiable those operational attributes. I changed the schema and create testcases to check pwdTPRValidFrom and pwdTPRExpireAt can be modified if an ACI grants the rights

Reply:
That's perfect! Thank you!